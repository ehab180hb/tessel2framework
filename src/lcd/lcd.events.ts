import { newMessage } from '../messages/messages.handler'
import { boardEvents, BoardEvents } from '../board/board.events'
import { fiveTypes } from '../typings/johnny-five'

interface UpdateScreenInput {
  row1?: string
  row2?: string
  lcd: fiveTypes.LCD
}

export const attachScreenUpdate = (updateScreenParams: UpdateScreenInput): void => {
  const emptyRow = ' '.repeat(16)
  const { row1 = emptyRow, row2 = emptyRow, lcd } = updateScreenParams
  const writeMsg = newMessage(lcd)
  writeMsg(0, row1)
  writeMsg(1, row2)
}

export const updateScreen = (updateScreenParams: UpdateScreenInput) => {
  boardEvents.emit(BoardEvents.UPDATE_SCREEN, updateScreenParams)
}
