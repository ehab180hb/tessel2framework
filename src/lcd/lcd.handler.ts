import * as five from 'johnny-five'
import { PIN } from '../board/board.model'
import { fiveTypes } from '../typings/johnny-five'

const pins: PIN[] = [PIN.a2, PIN.a3, PIN.a4, PIN.a5, PIN.a6, PIN.a7]

export const getLcd = (): fiveTypes.LCD => new five.LCD({ pins }) as any
