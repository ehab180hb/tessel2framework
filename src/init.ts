import { boardEvents, BoardEvents } from './board/board.events'
import { attachScreenUpdate } from './lcd/lcd.events'
import { attachButtonUpdate } from './button/button.events'

export const initComponents = () => {
  boardEvents.on(BoardEvents.UPDATE_SCREEN, attachScreenUpdate)
  boardEvents.on(BoardEvents.UPDATE_BUTTONS, attachButtonUpdate)
}
