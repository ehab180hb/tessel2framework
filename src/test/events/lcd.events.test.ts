import { attachScreenUpdate, updateScreen } from '../../lcd/lcd.events'
import { initComponents } from '../../init'
const lcd: any = {
  cursor: (x: any, y: any) => {},
  print: (x: any) => 'x',
}
describe('LCD events', () => {
  beforeEach(() => {
    lcd.cursor = jest.fn(() => lcd)
    lcd.print = jest.fn(() => lcd)
  })
  describe('attachScreenUpdate()', () => {
    test('works with default parameters', () => {
      attachScreenUpdate({ lcd })
      expect(lcd).toMatchSnapshot()
    })

    test('works with custom parameters', () => {
      attachScreenUpdate({ lcd, row1: 'hi', row2: 'hello' })
      expect(lcd).toMatchSnapshot()
    })
  })

  describe('updateButtons()', () => {
    test('Works with custom parameters', () => {
      initComponents()
      updateScreen({
        lcd,
        row1: `I'm green`,
        row2: 'therefore I exist',
      })
      expect(lcd).toMatchSnapshot()
    })
  })
})
