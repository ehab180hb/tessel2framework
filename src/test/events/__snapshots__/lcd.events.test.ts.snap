// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LCD events attachScreenUpdate() works with custom parameters 1`] = `
Object {
  "cursor": [MockFunction] {
    "calls": Array [
      Array [
        0,
        0,
      ],
      Array [
        0,
        0,
      ],
      Array [
        1,
        0,
      ],
      Array [
        1,
        0,
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
    ],
  },
  "print": [MockFunction] {
    "calls": Array [
      Array [
        "                ",
      ],
      Array [
        "hi",
      ],
      Array [
        "                ",
      ],
      Array [
        "hello",
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
    ],
  },
}
`;

exports[`LCD events attachScreenUpdate() works with default parameters 1`] = `
Object {
  "cursor": [MockFunction] {
    "calls": Array [
      Array [
        0,
        0,
      ],
      Array [
        0,
        0,
      ],
      Array [
        1,
        0,
      ],
      Array [
        1,
        0,
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
    ],
  },
  "print": [MockFunction] {
    "calls": Array [
      Array [
        "                ",
      ],
      Array [
        "                ",
      ],
      Array [
        "                ",
      ],
      Array [
        "                ",
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
    ],
  },
}
`;

exports[`LCD events updateButtons() Works with custom parameters 1`] = `
Object {
  "cursor": [MockFunction] {
    "calls": Array [
      Array [
        0,
        0,
      ],
      Array [
        0,
        0,
      ],
      Array [
        1,
        0,
      ],
      Array [
        1,
        0,
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
    ],
  },
  "print": [MockFunction] {
    "calls": Array [
      Array [
        "                ",
      ],
      Array [
        "I'm green",
      ],
      Array [
        "                ",
      ],
      Array [
        "therefore I exist",
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
      Object {
        "isThrow": false,
        "value": [Circular],
      },
    ],
  },
}
`;
